misc-utils/fincore.c

```diff
@@ -353,6 +355,41 @@ static int fincore_fd (struct fincore_control *ctl,
 	return mincore_fd(ctl, fd, st);
 }
 
+static void fincore_r(struct fincore_control *ctl,
+		      struct fincore_state *st)
+{
+	char * const paths[] = { (char *)st->name, NULL };
+	FTS *ftsp = fts_open(paths, FTS_PHYSICAL, NULL);
+	FTSENT *node;
+
+	if (!ftsp) {
+		warn(_("failed to open directory: %s"), st->name);
+		return;
+	}
+
+	while ((node = fts_read(ftsp)) != NULL) {
+		if (node->fts_info == FTS_F) {
+			int fd = open(node->fts_accpath, O_RDONLY);
+
+			if (fd >= 0) {
+				struct fincore_state st_child = {
+					.name = node->fts_path,
+				};
+
+				fincore_fd(ctl, fd, &st_child);
+				add_output_data(ctl, &st_child);
```

Comment:
It is better to call add_output_data only if fincore_fd returns 0.

Reply:
I'm thinking about, if --recursive, call fts_open directly in main and reusing fincore_name(), so we will still handle block devices